### PR TITLE
ar71xx: add support for 8devices Lima board

### PR DESCRIPTION
**Pull request by karoiz**
Lima is a module for the (IoE) internet of everything applications,
based on Qualcomm/Atheros QCA4531.

Specification:

- 650/600/216 MHz (CPU/DDR/AHB)
- 64 MB of RAM (DDR2)
- 32 MB of FLASH
- 2T2R 2.4 GHz
- 2x 10/100 Mbps Ethernet
- USB 2.0 Host
- PCIe
- UART for serial console
- 14x GPIO

Flash instruction:
1. Download lede-ar71xx-generic-lima-squashfs-sysupgrade.bin
and rename it to 8dev_recovery.bin
2. Binary file copy to USB stick and plug into slot USB
3. Press S2 (reset) button
4. Press and hold the S1 button by 5 seconds
5. Wait ~40 seconds to complete recovery

Tested on:

- Lima development kit

Signed-off-by: Karol Dudek <karoiz@sli.pl>

### CODE REVIEW

File: target/linux/ar71xx/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom [hunk 1]
```diff
@@ -58,6 +58,9 @@ case "$FIRMWARE" in
 		ath9k_eeprom_extract "art" 4096 2048
 		ath9k_patch_firmware_mac $(macaddr_add $(mtd_get_mac_binary art 0) +1)
 		;;
+	lima)
```
Comment (pepe2k) [inline target/linux/ar71xx/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom]: This is redundant here, please use `ath79_register_wmac()` in mach-lima.c with correct offset.

File: target/linux/ar71xx/base-files/lib/preinit/05_set_iface_mac_ar71xx [hunk 2]
```diff
@@ -34,6 +34,12 @@ preinit_set_mac_address() {
 		dir-615-i1)
 			fetch_mac_from_mtd nvram sys_lan_mac sys_wan_mac
 			;;
+		lima)
```
Comment (pepe2k) [inline target/linux/ar71xx/base-files/lib/preinit/05_set_iface_mac_ar71xx]: Same here, please use `ath79_init_mac()` in mach-lima.c.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-lima.c [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+#define LIMA_GPIO_BTN_RST_DEFAULT	16
+#define LIMA_KEYS_POLL_INTERVAL		20	/* msecs */
+#define LIMA_KEYS_DEBOUNCE_INTERVAL	(3 * LIMA_KEYS_POLL_INTERVAL)
+
+#define LIMA_ETH_PHYS	(BIT(0) | BIT(1))
+
+static struct gpio_keys_button lima_gpio_keys[] __initdata = {
+	{
+		.desc		= "button0",
```
Comment (pepe2k) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-lima.c]: You mentioned `S1` button in commit message, why then `button0` here?

**Comment by karoiz** [comment]
Thank you for your tips. I have implemented changes in source code.

File: target/linux/ar71xx/image/generic.mk [hunk 4]
```diff
@@ -161,6 +161,17 @@ define Device/gl-mifi
...
 TARGET_DEVICES += gl-mifi
 
+define Device/lima
+  DEVICE_TITLE := Lima board from 8Devices
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2
+  BOARDNAME = LIMA
+  BLOCKSIZE = 256k
+  IMAGE_SIZE = 15616k
+  MTDPARTS = spi0.0:256k(u-boot)ro,256k(u-boot-env)ro,256k(art)ro,-(firmware)
+  IMAGE/sysupgrade.bin = append-kernel $$(BLOCKSIZE) | append-rootfs | pad-rootfs | check-size $$$$(IMAGE_SIZE)
```
Comment (pepe2k) [inline target/linux/ar71xx/image/generic.mk]: This line doesn't look correct for me (this part: `append-kernel $$(BLOCKSIZE)` is wrong).
I think you can just drop the whole line and relay on [default definition](https://github.com/lede-project/source/blob/master/target/linux/ar71xx/image/Makefile#L110).

File: target/linux/ar71xx/image/generic.mk [hunk 5]
```diff
@@ -161,6 +161,17 @@ define Device/gl-mifi
 endef
 TARGET_DEVICES += gl-mifi
 
+define Device/lima
+  DEVICE_TITLE := Lima board from 8Devices
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2
+  BOARDNAME = LIMA
+  BLOCKSIZE = 256k
```
Comment (pepe2k) [inline target/linux/ar71xx/image/generic.mk]: Can you please explain 256k instead of default 64k block size?

**Comment by karoiz** [comment]
In old repo OpenWRT had a problem with large flash and block size 64k. Now it is not leaking, corrections applied. On the new repo LEDE it's not problem, I used the size of 64k.

**Comment by pepe2k** [comment]
@karoiz, thanks for all the changes and your patch!
@blogic or @mkresin, this one looks OK for me, could you also take a look and merge it if there are no more comments?

Cheers,
Piotr

**Comment by blogic** [comment]
merge, Thanks !

**Pull request closed without merging**